lib/ansible/runner/credential.rb

```diff
@@ -0,0 +1,56 @@
+module Ansible
+  class Runner
+    class Credential
+      attr_reader :auth, :base_dir
+
+      def self.new(authentication_id, base_dir)
+        auth_type = Authentication.find(authentication_id).type
+        self == Ansible::Runner::Credential ? detect_credential_type(auth_type).new(authentication_id, base_dir) : super
+      end
+
+      def self.detect_credential_type(auth_type)
+        subclasses.index_by(&:auth_type)[auth_type] || Ansible::Runner::GenericCredential
+      end
+
+      def initialize(authentication_id, base_dir)
+        @auth     = Authentication.find(authentication_id)
+        @base_dir = base_dir
+      end
+
+      def command_line
+        {}
+      end
+
+      def env_vars
+        {}
+      end
+
+      def extra_vars
+        {}
+      end
+
+      def write_password_file
+      end
+
+      private
+
+      def password_file
+        File.join(env_dir, "passwords")
+      end
+
+      def ssh_key_file
+        File.join(env_dir, "ssh_key")
+      end
+
+      def env_dir
+        File.join(base_dir, "env")
+      end
+
+      def ensure_env_dir
```

Comment:
You could even "ensure env dir" in the initialize method if you want to keep this method.  I just think having to call `ensure_env_dir` in the sub class is a code smell.

Reply:
That's fair. I was trying to avoid creating it if we didn't need it, but I don't think we really care if it exists or not. Also, this may not belong in credentials as all. Maybe we build out the whole directory structure (`env`, `inventory`, `project`, `roles`) at the top of `Ansible::Runner.run_via_cli`?

Then we can get rid of stuff like [this](https://github.com/ManageIQ/manageiq/blob/master/lib/ansible/runner.rb#L244) and the duplication [here](https://github.com/ManageIQ/manageiq/blob/master/lib/ansible/runner.rb#L265)

Reply:
Ideally we could do this as a part of initializing an `Ansible::Runner` instance, but alas ... class methods and I don't feel like refactoring that much right now.

Reply:
So I ended up just moving this to `initialize` because moving everything to `Ansible::Runner` made specs annoying and coupled things together too tightly for my liking. (The credential classes were assuming that someone would have created the directories they needed south of `base_dir`).